src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+
+    /// The absolute path to the supergraph configuration file.
+    #[serde(skip_serializing)]
+    #[arg(long = "supergraph-config")]
+    supergraph_yaml: Option<Utf8PathBuf>,
+}
+
+impl Lsp {
+    pub async fn run(&self, client_config: StudioClientConfig) -> RoverResult<RoverOutput> {
+        self.opts
+            .plugin_opts
+            .prompt_for_license_accept(&client_config)?;
+
+        run_lsp(client_config, self.opts.clone()).await?;
+        Ok(RoverOutput::EmptySuccess)
+    }
+}
+
+async fn run_lsp(client_config: StudioClientConfig, lsp_opts: LspOpts) -> RoverResult<()> {
+    //TODO: Check this error handling is right i.e. if we don't get a supergraph.yaml passed
+    // this should fail rather than doing something else.
+    let supergraph_yaml_path = lsp_opts
+        .supergraph_yaml
+        .as_ref()
+        .and_then(|path| {
+            if path.is_relative() {
+                Some(
+                    Utf8PathBuf::try_from(std::env::current_dir().ok()?)
+                        .ok()?
+                        .join(path),
+                )
+            } else {
+                Some(path.clone())
+            }
+        })
+        .ok_or_else(|| anyhow!("Could not find supergraph.yaml file."))?;
+    let supergraph_content_root = supergraph_yaml_path.parent().unwrap().to_path_buf();
+
+    let studio_client = client_config.get_authenticated_client(&lsp_opts.plugin_opts.profile)?;
+
+    // Get the SupergraphConfig in a form we can use
+    let supergraph_config = SupergraphConfigResolver::default()
+        .load_remote_subgraphs(&studio_client, None)
+        .await?
+        .load_from_file_descriptor(
+            &mut stdin(),
+            Some(&FileDescriptorType::File(supergraph_yaml_path.clone())),
+        )?;
+    let lazily_resolved_supergraph_config = supergraph_config
+        .lazily_resolve_subgraphs(&supergraph_content_root)
+        .await?;
+
+    // Build the service to spin up the language server
+    let (service, socket, _receiver) = ApolloLanguageServer::build_service(
+        Config {
+            root_uri: supergraph_content_root.to_string(),
+            enable_auto_composition: false,
+            force_federation: false,
+            disable_telemetry: false,
+        },
+        HashMap::from_iter(
+            lazily_resolved_supergraph_config
+                .subgraphs()
+                .iter()
+                .map(|(a, b)| (a.to_string(), b.schema().clone())),
+        ),
+    );
+
+    let language_server = service.inner().clone();
+
+    // Spawn a separate thread to handle composition and passing that data to the language server
+    tokio::spawn(async move {
+        // Create a supergraph binary
+        // TODO: Check defaulting behaviour here and see if we need to centralise
+        let federation_version = lazily_resolved_supergraph_config
+            .federation_version()
+            .clone()
+            .unwrap_or(FederationVersion::LatestFedTwo);
```

Comment:
Have we centralised the behaviour on detecting the Federation Version? I feel like we definitely should if we haven't already

Reply:
oh, came across this in the source while looking at something else: [panic when calling composition_did_update on a non-supergraph](https://docs.rs/apollo-language-server/0.3.0/src/apollo_language_server/server.rs.html#411), so maybe requiring the supergraph config yaml is the strat!

this also raises a question about how panics are being handled; what happens in rover when the LSP server takes a dirt nap from some internal panic?

Reply:
for posterity: I believe @dotdat is working on this?